service/datastore/client-rest/src/main/java/org/eclipse/kapua/service/datastore/client/rest/RestDatastoreClient.java

```diff
@@ -0,0 +1,641 @@
+                for (JsonNode item : items) {
+                    JsonNode itemNode = item.get(KEY_UPDATE);
+                    if (itemNode != null) {
+                        TextNode idNode = ((TextNode) itemNode.get(KEY_DOC_ID));
+                        String metricId = null;
+                        if (idNode != null) {
+                            metricId = idNode.asText();
+                        }
+                        String indexName = ((TextNode) itemNode.get(KEY_DOC_INDEX)).asText();
+                        String typeName = ((TextNode) itemNode.get(KEY_DOC_TYPE)).asText();
+                        int responseCode = ((NumericNode) itemNode.get(KEY_STATUS)).asInt();
+                        if (!isRequestSuccessful(responseCode)) {
+                            TextNode failureNode = ((TextNode) itemNode.get(KEY_RESULT));
+                            String failureMessage = MSG_EMPTY_ERROR;
+                            if (failureNode != null) {
+                                failureMessage = failureNode.asText();
+                            }
+                            bulkResponse.add(new UpdateResponse(metricId, new TypeDescriptor(indexName, typeName), failureMessage));
+                            logger.info("Upsert failed [{}, {}, {}]", new Object[] { indexName, typeName, failureMessage });
+                            continue;
+                        }
+                        bulkResponse.add(new UpdateResponse(metricId, new TypeDescriptor(indexName, typeName)));
+                        logger.debug("Upsert on channel metric succesfully executed [{}.{}, {}]", new Object[] { indexName, typeName, metricId });
+                    } else {
+                        throw new ClientException(ClientErrorCodes.ACTION_ERROR, "Unexpected action response");
+                    }
+                }
+                return bulkResponse;
+            } else {
+                throw new ClientException(ClientErrorCodes.ACTION_ERROR, updateResponse.getStatusLine().getReasonPhrase());
+            }
+        } catch (IOException e) {
+            throw new ClientException(ClientErrorCodes.ACTION_ERROR, e, e.getLocalizedMessage());
+        }
+    }
+
+    @Override
+    public <T> T find(TypeDescriptor typeDescriptor, Object query, Class<T> clazz) throws ClientException {
+        checkClient();
+        ResultList<T> result = query(typeDescriptor, query, clazz);
+        if (result.getTotalCount() == 0) {
+            return null;
+        } else {
+            return result.getResult().get(0);
+        }
+    }
+
+    @Override
+    public <T> ResultList<T> query(TypeDescriptor typeDescriptor, Object query, Class<T> clazz) throws ClientException {
+        checkClient();
+        JsonNode queryMap = queryConverter.convertQuery(query);
+        Object queryFetchStyle = queryConverter.getFetchStyle(query);
+        logger.debug("Query - converted query: '{}'", queryMap.toString());
```

Comment:
Remove `.toString()` is already called `.toString()` for each parameter passed as argument.